File: maven-leetcode/src/main/java/leetcode/Num53.java
package leetcode;

public class Num53 {
    class Solution {
        /**
         * 这题是最大子序列和的问题
         * 采用动态规划算法，令dp[i]表示以i结尾的最大子序列和
         * 则dp[i + 1] = max{dp[i] + nums[i],nums[i]} */
        public int maxSubArray(int[] nums) {
            int [] dp = new int[nums.length];
            int max = nums[0];
            dp[0] = nums[0];
            for (int i = 1; i < dp.length; i++){
                if(dp[i-1] > 0) {
                    dp[i] = dp[i - 1] + nums[i];
                } else {
                    dp[i] = nums[i];
                }
                if (dp[i] > max) {
                    max = dp[i];
                }
            }
            return max;
        }
    }
}

File: learnPython/num53.py
class Solution:
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        size = len(nums)
        if size == 0:
            return 0
        max = 0
        dp = list(range(size+1))
        for i,val in enumerate(nums):
            if(dp[i] > 0):
                dp[i+1]=dp[i]+val
            else:
                dp[i+1] = val
            if dp[i+1]>max:
                max = dp[i+1]
        return max
if __name__=='__main__':
    s = Solution()
    nums = [-2,1,-3,4,-1,2,1,-5,4]
    print(s.maxSubArray(nums))